### PR TITLE
traceloop: use custom resources

### PR DESCRIPTION
**Pull request by alban**
# traceloop: use custom resources

Traceloop is no longer started automatically and is controlled by the Trace resource.

## How to use

It can be started manually with a Trace custom resource on each node:
```
$ kubectl apply -f ./pkg/resources/samples/trace-traceloop.yaml
$ kubectl annotate -n gadget trace/traceloop gadget.kinvolk.io/operation=start
```
The CLI is updated to support start and stop operations:
```
$ kubectl gadget traceloop start
$ kubectl gadget traceloop stop
```

## Testing done

- Tested with the operations above on AKS

TODO:
- [x] Rebase on top of #333 and use less arguments on `GenericTraceCommand()`
- [x] Update traceloop guide
- [x] Don't allow several instances of the traceloop gadget on the same node
- [x] Delete the `traceloop.kinvolk.io/state` annotation on the gadget pods when the trace is deleted

### CODE REVIEW

**Comment by mauriciovasquezbernal** [comment]
Will hold until https://github.com/kinvolk/inspektor-gadget/pull/357 is merged. Otherwise there are going to be so many conflicts to solve.

**Comment by mauriciovasquezbernal** [comment]
I updated the PR. I decided to handle the annotation removal directly in traceloop https://github.com/kinvolk/traceloop/pull/65. Regarding my comment about `I have to wait a bit after starting traceloop to be able to list the traces:`, I think we can solve it as part of https://github.com/kinvolk/inspektor-gadget/issues/371.

**Comment by mauriciovasquezbernal** [comment]
This is ready for review. @alban can you review it? otherwise @blanquicet can you?

File: cmd/kubectl-gadget/traceloop.go
Comment (alban) [inline cmd/kubectl-gadget/traceloop.go:168]: No need for `!` :)
```suggestion
		err = fmt.Errorf("Traceloop is not running on any node. Please start it with 'kubectl gadget traceloop start'.")
```

File: cmd/kubectl-gadget/traceloop.go
Comment (alban) [inline cmd/kubectl-gadget/traceloop.go:196]: TODO for another PR: what if traceloop is started but then a new node is added to the cluster? It would not be possible to start traceloop there without stopping it on all nodes first.

This kind of problem should be solved with a GlobalTrace CR.

File: cmd/kubectl-gadget/utils/trace.go
Comment (alban) [inline cmd/kubectl-gadget/utils/trace.go:738]: Is it possible to do this with a field selector instead of using the label?

```
FieldSelector: fmt.Sprintf("spec.gadget=%s", gadget)
```
We have an example of that, but for another resource:
https://github.com/kinvolk/inspektor-gadget/blob/f5e0b23afbf4659c0ed8a4b60d7fc63d6c7ebac1/cmd/kubectl-gadget/utils/exec.go#L50

**Review by alban — COMMENTED** [review]
Some suggestions but LGTM.
Thanks!
I cannot approve the PR since I created it.

File: cmd/kubectl-gadget/utils/trace.go
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/utils/trace.go:738]: It didn't work: `field label not supported: spec.gadget`. Kubernetes objects have a [`ToSelectableFields`](https://github.com/kubernetes/kubernetes/blob/c2882f314218af489267b7163b2ca09c79ffb00a/pkg/registry/core/pod/strategy.go#L306) function to set the fields that could be used in the lookup. Do you have any idea where can be implement them for the trace resource?

Comment (alban) [reply cmd/kubectl-gadget/utils/trace.go]: It seems that for custom resources, the field selector can only work with the namespace and name but not arbitrary fields.

https://github.com/kubernetes/kubernetes/issues/79233
https://github.com/kubernetes/kubernetes/issues/53459

**Comment by alban** [comment]
LGTM if the CI is green

**Comment by mauriciovasquezbernal** [comment]
> LGTM if the CI is green

Ah, seems like I introduced a deadlock... I'll investigate.

File: cmd/kubectl-gadget/utils/trace.go
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/utils/trace.go]: Thanks for the clarification!

**Review by mauriciovasquezbernal — APPROVED** [review]
Alban was the original author so he can't approve this PR. I'm approving it as he already gave a LGTM!

**Pull request merged**
